File: src/main/java/project/service/CustomerService.java
package project.service;

import project.entity.Customer;
import org.springframework.http.ResponseEntity;

import java.util.List;
import java.util.Map;

public interface CustomerService {
    ResponseEntity<String> addNewCustomer(Map<String, String> requestMap);
    ResponseEntity<List<Customer>> getAllCustomers(String filterValue);
    ResponseEntity<String> updateCustomer(Map<String, String> requestMap);
    ResponseEntity<Customer> getById(Integer id);
    ResponseEntity<String> deleteCustomer(Integer id);
}

File: src/front/train-booking-frontend/src/services/customer.service.ts

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../environments/environment';

export interface Customer {
  id?: number;
  name: string;
}

export interface ApiResponse {
  message: string;
}

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  private apiUrl = `${environment.apiUrl}/customer`;

  constructor(private http: HttpClient) {}

getAllCustomers(filterValue?: string): Observable<Customer[]> {
  const url = filterValue 
    ? `${this.apiUrl}/get?filterValue=${filterValue}` 
    : `${this.apiUrl}/get`;
  return this.http.get<Customer[]>(url);
}

  getCustomerById(id: number): Observable<Customer> {
    return this.http.get<Customer>(`${this.apiUrl}/getById/${id}`);
  }

  addCustomer(customer: Customer): Observable<ApiResponse> {
    return this.http.post<ApiResponse>(`${this.apiUrl}/add`, customer);
  }

  updateCustomer(customer: Customer): Observable<ApiResponse> {
    return this.http.put<ApiResponse>(`${this.apiUrl}/update`, customer);
  }

  deleteCustomer(id: number): Observable<ApiResponse> {
    return this.http.delete<ApiResponse>(`${this.apiUrl}/delete/${id}`);
  }
}